(PDB-2122) New 'root' single query endpoint and 'from' operator

kbarber opened the pull request:
This patch introduces the new 'root' endpoint that allows for querying any
entity inside a single query by providing a context query.

Here I've included this new feature all the way back to produce-streaming-body,
stream-query-result and the query API. This patch means that all the HTTP
endpoints are required to wrap their queries in from, so that this concept
of a single query for all endpoints exists all the way back to the query engine
API.

For cleanup, I've modified the produce-streaming-body and stream-query-results
to take an options map instead of a list of ordered arguments as we seem to change
this frequently and some validation is desired. Plus this simplifies argument
passing as most of the data is already available in `globals`.

As a side-effect of adding the from operator, I've added a new format for
subqueries that uses the same syntax at this level, as it does on the top-level.
This is provided for consistency, and it will hopefully remove the need for
the select_<entity> operator which was always a bit special.

Signed-off-by: Ken Barber <ken@bob.sh>

kbarber commented:
There is a bit of a chicken and egg scenario for merge tests to pass for this PR. When we merge this pe-puppetdb-extensions will fail, so we'll need this PR also merged: https://github.com/puppetlabs/pe-puppetdb-extensions/pull/144

kbarber commented on documentation/_puppetdb_nav.html:
```diff
@@ -60,6 +60,7 @@ <h2 id="puppetdb">PuppetDB {{ puppetdb_version }}</h2>
       <li><a href="/puppetdb/{{ puppetdb_version }}/api/query/v4/query.html">Query Structure</a></li>
       <li><a href="/puppetdb/{{ puppetdb_version }}/api/query/v4/operators.html">Available Operators</a></li>
       <li><a href="/puppetdb/{{ puppetdb_version }}/api/query/v4/paging.html">Query Paging</a></li>
+      <li><a href="/puppetdb/{{ puppetdb_version }}/api/query/v4/root.html">Root Endpoint (Experimental)</a></li>
```
I couldn't think of a better name for the endpoint, ideas welcome.

ajroetker replied:
Would using the route `/pdb/query/v4` be crazy?
If so what about `/pdb/query/v4/context`  or `/pdb/query/v4/from` to match it with the operator?

kbarber replied:
It actually is /pdb/query/v4 ... I'm pointing out 'Root Endpoint' here as this is just the name in documentation. I could have called it 'Index Endpoint' instead, and make the docs appear in index.html, but I just wasn't sure.

ajroetker replied:
Oh nice sorry, misunderstood. Is there a way we could redirect `/puppetdb/{{ puppetdb_version }}/api/query/v4` to `/puppetdb/{{ puppetdb_version }}/api/query/v4/root.html`? Otherwise the name is :+1:  from me.

kbarber replied:
Well, I can just make it index.html. Saves any hassle that way.

kbarber commented on src/puppetlabs/puppetdb/http/aggregate_event_counts.clj:
```diff
@@ -5,15 +5,16 @@
 (defn routes
   [version optional-handlers]
-  (let [param-spec {:required ["summarize_by"]
+  (let [handlers (cons (partial http-q/restrict-query-to-entity "aggregate_event_counts")
```
endpoints need to wrap their queries in a from as you can see here now, since the from is expected in the query when using the query engine functions.

kbarber commented on src/puppetlabs/puppetdb/http/query.clj:
```diff
@@ -290,31 +299,14 @@
-      (extract-query param-spec)
-      (apply comp
-             (fn [{:keys [params globals puppetdb-query]}]
-               (warn-experimental entity (:warn-experimental globals))
```
I pushed experimental warning on entities into the query engine, as it feels like a 'lower-down' problem.

kbarber commented on src/puppetlabs/puppetdb/http/server.clj:
```diff
@@ -11,18 +11,6 @@
             [net.cgrand.moustache :refer [app]]
             [ring.util.response :as rr]))
 
-(defn deprecated-app
```
I've re-used experimental-warning elsewhere, so I moved these both to http.clj for general use.

senior replied:
:+1:

kbarber commented on src/puppetlabs/puppetdb/query_eng.clj:
```diff
@@ -101,55 +106,77 @@
       (munge-fn version url-prefix))))
 
-(defn stream-query-result
+(def query-options-schema
```
produce-streaming-body and stream-query-results now takes an options hash that gets validated and has defaults. We keep moving around these parameters and this feels like a step in a better direction to manage these, at least produce-streaming-body has a contract with pe-puppetdb-extensions for example.

kbarber commented on src/puppetlabs/puppetdb/query_eng/engine.clj:
```diff
@@ -1539,6 +1574,39 @@
+                "The %s entity is experimental and may be altered or removed in the future."
+                (name entity)))))
+
+(defn parse-query-context
```
This magic extracts the 'from' and returns the remaining query and entity. In a way its a compatibility layer, later on we want to push the 'from' further down into the query engine, but munging still happens at the query_engine.clj level, so we still need this extraction for now ...

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/297/
Test PASSed.

ajroetker commented on src/puppetlabs/puppetdb/http/aggregate_event_counts.clj:
```diff
@@ -5,15 +5,16 @@
                                "distinct_end_time"]}
-        query-route #(apply (partial http-q/query-route :aggregate-event-counts
-                                     version param-spec) %)]
+        query-route #(apply (partial http-q/query-route version param-spec) %)]
```
Should we just make a wrapper around `http-q/query-route` which takes a list like this interface expects?

kbarber replied:
Good thinking, doing that now. Plus, I've made a query-route-from wrapper that automatically puts in the restriction for you, saves on some loc at least :-).

ajroetker commented on src/puppetlabs/puppetdb/query_eng.clj:
```diff
@@ -56,7 +58,10 @@
   [entity]
-  (get-in @entity-fn-idx [entity :munge]))
+  (let [munge-result (get-in @entity-fn-idx [entity :munge])]
+    (if (nil? munge-result)
```
```clojure
(if-let [munge-result (...)]
 munge-result
 (throw ...))
```

kbarber replied:
Ta :-).

ajroetker commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -318,10 +318,10 @@
-               url-prefix @(get (service-context this) :url-prefix)]
-           (qeng/stream-query-result query-obj version query-expr paging-options db url-prefix row-callback-fn))))
+  (query [this version query-expr paging-options row-callback-fn]
+         (let [query-options (-> (select-keys (get (service-context this) :shared-globals) [:scf-read-db :warn-experimental])
```
I might pull out the `(get (service-context this) :shared-gloabls)` to be the first argument of the `->`

ajroetker replied:
Or even pull out the `(service-context this)` to a named variable in the `let`

kbarber replied:
Cool.

ajroetker commented on src/puppetlabs/puppetdb/http/edges.clj:
```diff
@@ -12,11 +12,10 @@
      (app
        [""]
-       (query-route handlers)))))
+       (http-q/query-route-from "edges" version param-spec handlers)))))
```
Nice :dancer:, so much easier to read!

kbarber replied:
Agreed

ajroetker commented on src/puppetlabs/puppetdb/http/v4.clj:
```diff
@@ -13,12 +14,24 @@
+  (fn [request]
+    (http/experimental-warning
+     (index/index-app version)
+     "The index endpoint is experimental"
```
Should call this `root` to match the docs?

kbarber replied:
Yep.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/301/
Test PASSed.

ajroetker commented:
Tested out bunch of queries using this patch and it all lgtm. One note that is a preexisting issue, invalid operators will fail silently. For instance I tried `-d 'query="["from","reports",["exadsacnsldc","certname"]]' and PDB returned the unextracted reports.

ajroetker commented:
Also with the addition of this endpoint, it would be really nice if we had some kind of discovery endpoint or doc which graphed out all the valid entities for the `from` operators. Something like `/pdb/query` returning:
```json
{"/v4" : ["reports","catalogs",...,"report_logs"]} 
```
We could even end up adding the valid implicit `subquery` relationships or some such.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/302/
Test PASSed.

kbarber commented:
@ajroetker for the discovery/doc side of things: I think discovery is someone we can circle back on, its a complex topic - let me ponder it some more. I do notice however, I fail to link to the available entities properly in docs, I've got some ideas on this and I'll submit something soon.

I'll take a quick look into the operator issue, if its easy I'll fix otherwise I'll create a ticket.

kbarber commented:
@ajroetker I've added the new entities.markdown doc now and updated some other docs related to this problem, that splits out the available entities. Seems we were using it all over the place, so I've modded those places to reference this, actually simplifying our docs somewhat. I think _this_ is the place we should have our relationship diagram also ... which somewhat answers that question I've had in the back of my head also.

kbarber commented on documentation/api/query/v4/entities.markdown:
```diff
@@ -0,0 +1,53 @@
+---
+title: "PuppetDB 3.2 » API » v4 » Entities"
```
I think this is where our relationship diagram could eventually live.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/303/
Test FAILed.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/304/
Test FAILed.

kbarber commented:
Those last two failures are because I cancelled those tests ^^. I did this on purpose, to only leave the last one running.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/305/
Test PASSed.

ajroetker commented on documentation/api/query/v4/entities.markdown:
```diff
@@ -0,0 +1,53 @@
+direct querying.
+
+The entity name is utilised within queries themselves, in particular within the [`from`][from]
+and [`subquery`][subquery] operators.
```
This is great, I've been wanting this for a long time.

kbarber replied:
Yeah, its basically replacing select_<entity> ... and providing upper and lower-level syntax consistency in doing so. My decision to move the 'from' part to the outside though, is more that even in things like SQL, the table context is the mandatory part, extract is not mandatory in principle (ie. select * from foo).

ajroetker commented on src/puppetlabs/puppetdb/http/query.clj:
```diff
@@ -290,31 +299,25 @@
+  ([entity version param-spec optional-handlers]
+   (let [handlers (cons (partial restrict-query-to-entity entity)
+                        optional-handlers)]
+     (query-route version param-spec handlers))))
```
Any way that query-route and query-route-from could be merged, or could the `([version param-spec] (query-route version param-spec [identity]))` case of query-route be removed, just a little bothered by the duplication of the optional-handlers defaulting interface?

kbarber replied:
Done, second one.

ajroetker commented on src/puppetlabs/puppetdb/http/nodes.clj:
```diff
@@ -13,14 +13,12 @@
-                                         api-version
-                                         ["=" "certname" node]
+                (eng/stream-query-result api-version
+                                         ["from" "nodes" ["=" "certname" node]]
```
Just making a note that anything that runs `stream-query-result` directly might not handle database streaming errors correctly, and also won't pretty-print.

ajroetker replied:
A pre-existing problem*

senior replied:
@ajroetker can you check on this and create a ticket if you find anything?

ajroetker commented on src/puppetlabs/puppetdb/http/index.clj:
```diff
@@ -0,0 +1,18 @@
+  [version optional-handlers]
+  (let [param-spec {:optional paging/query-params
+                    :required ["query"]}]
+    (app
```
I'd prefer us using compojure instead of moustache, like we do for `src/puppetlabs/puppetdb/metrics.clj`.

kbarber replied:
All the other query http endpoints use this methodology, this feels like a refactor wishlist for later.

senior replied:
Meh, I think new endpoints have been compojure. That's certainly where we want to go. I think this would be pretty simple to make compojure as well, just adding work for ourselves later if we don't.

kbarber replied:
Its more a matter of, 'might as well do them all' since all the existing patterns are the same today, instead of progressing in piecemeal and having the endpoints look different, yet again.

senior replied:
Discussed with the group, it does add some work to move to compojure (or some other lib) but we're going to proceed with it and get the routing lib and refactor in the next sprint or two.

ajroetker commented:
Just some small notes, otherwise definitely :+1: again!

kbarber commented:
@ajroetker new patch up again.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/316/
Test PASSed.

ajroetker commented on documentation/api/query/v4/operators.markdown:
```diff
@@ -176,6 +178,30 @@ To get the average uptime for your nodes,
+
+Setting context in a query allows you to choose the entity you are querying
+on. This augments the endpoint support we have today, whereby the endpoint
+decides the context by allowing a single endpoint to contain all the information
```
I think you can cut off this sentence and just say "whereby the endpoint decides the context. For example `/pdb/query/v4/nodes` sets the context of the query to `nodes`.

ajroetker commented on documentation/api/query/v4/operators.markdown:
```diff
@@ -176,6 +178,30 @@ To get the average uptime for your nodes,
+
+### `from`
+
+The `from` operator allows one to choose the entity that you want to query and
```
Could `entity` here have a link to the "entities" doc? Channeling my inner Wyatt w.r.t. entity.

ajroetker commented on documentation/api/query/v4/operators.markdown:
```diff
@@ -176,6 +178,30 @@ To get the average uptime for your nodes,
+using the [`in` operator](#subquery-operators).
+
+When providing a query to a particular endpoint such as `/nodes` the endpoint decides
+the context. With the new [root] endpoint, a query must contain the context.
```
> When querying a particular endpoint, such as `/nodes`, the endpoint provides the context for the query. While querying the [root] endpoint requires specifying a context explicitly.

ajroetker commented on documentation/api/query/v4/operators.markdown:
```diff
@@ -197,20 +223,10 @@ related entity and the query to use:
-* [`reports`][reports]
-* [`resources`][resources]
+The [`<ENTITY>`][entities] is the particular entity you are subquerying on, however not
+all entities are relateable to other entities, as not every relationship makes sense.
```
> not all entities are implicitly relatable to all other entities

ajroetker commented on documentation/api/query/v4/operators.markdown:
```diff
@@ -395,3 +431,50 @@ For the previous query, we also allow the shorthand
+              ["=", "type", "Class"],
+              ["=", "title", "Apache"]]]]]]
+
+This queries on the `/facts` endpoint filters for uptime_hours of all nodes with
```
> Executing this query on the `/facts` endpoint would filter for `uptime_hours` for all nodes with `facts_environment` set to `production`:

ajroetker commented on documentation/api/query/v4/operators.markdown:
```diff
@@ -395,3 +431,50 @@ For the previous query, we also allow the shorthand
+            ["=", "facts_environment", "production"]]]]]
+
+To find node information for a host that has a macaddress of `aa:bb:cc:dd:ee:00` as
+its first macaddress on the interface `eth0`, you could use this query on '/nodes':
```
'/nodes' -> `/nodes`

ajroetker commented on documentation/api/query/v4/query.markdown:
```diff
@@ -35,6 +36,12 @@ That is, most queries will look like a GET request to a URL that resembles the f
 
     https://puppetdb:8081/pdb/query/v4/<ENDPOINT>?query=<QUERY STRING>
 
+Alternatively, you can provide the entity instead of using the `<ENDPOINT>` suffix with the following:
```
> Alternatively, you can provide the context/entity in a query, instead of using the `<ENPOINT>` suffix, with the following:

ajroetker commented:
Proofread the documentation updates @kbarber, pretty minor stuff. I've been through the code and docs as much as I can! If someone else /cc @rbrw @mullr @senior wants to just give a separate proofread through the code and docs, I'm ready to merge this in.

kbarber commented:
@ajroetker updated from your review.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/318/
Test PASSed.

ajroetker commented:
@kbarber I noticed that `pretty=true` is not a valid query param to this endpoint. It might be nice to add that as an option.

senior commented on documentation/api/query/v4/index.markdown:
```diff
@@ -0,0 +1,66 @@
+a query with a `from` is required to choose the entity to query for. For
+general info about queries, see [the page on query structure.][query]
+
+If a query parameter is not provided, all results will be returned.
```
Is this a copy and past error? Above talks about requiring a query parameter.

kbarber replied:
Yes it is.

senior commented on documentation/api/query/v4/index.markdown:
```diff
@@ -0,0 +1,66 @@
+### Response Format
+
+The response will be in `application/json`, and will contain a list of JSON
+object results based on the query provided.
```
I think we need to say more about the response format here. Maybe with links that go to the from entity and the correct response format?

kbarber replied:
I'll link to the entities page a little better. I'm avoiding repeating myself here, since that list will become out of date. Sometimes, we need the ability to insert sub-documents, rather than link to them I think.

senior commented on documentation/api/query/v4/index.markdown:
```diff
@@ -0,0 +1,66 @@
+other endpoints, the [entity][entities] must be supplied using a query with the [`from`][from]
+operator.
+
+### URL Parameters
```
We should also have links here to the entities and their available query fields.

kbarber replied:
Same as above, I'll add more linking to the entities page.

senior commented on src/puppetlabs/puppetdb/http/fact_names.clj:
```diff
@@ -14,27 +14,26 @@
-               (produce-streaming-body
-                 entity
+  [version param-spec & optional-handlers]
+  (let [handlers (cons (partial http-q/restrict-query-to-entity "fact_names")
```
Why not omit the let (and the cons) and just add the partial and optional handlers in to the comp below?

kbarber replied:
+1

senior commented on src/puppetlabs/puppetdb/http/query.clj:
```diff
@@ -179,6 +175,19 @@
   (restrict-query ["=" "title" title]
                   req))
 
+(defn wrap-in-from
```
fwiw I expected to see this function use an in clause and a from clause until seeing the implementation.

kbarber replied:
or reading the note below? I can change it to wrap-with-from or something.

senior commented on src/puppetlabs/puppetdb/query_eng.clj:
```diff
@@ -101,55 +105,77 @@
+    options :- query-options-schema
+    row-fn]
+   (let [{:keys [scf-read-db url-prefix warn-experimental pretty-print]
+          :or {warn-experimental true
```
I'm not sure I'm understanding `warn-experimental` and that it defaults to true. Is this saying by default we're output a warning for any call to stream-query-result that doesn't set it to false?

kbarber replied:
The concept is that we didn't want to warn that those few "distinct" entities were now experimental in PE I believe. By default we always throw warnings for experimental entities, unless its switched off like in PE. This applies to event_counts and aggregate_event_counts.

senior commented on src/puppetlabs/puppetdb/query_eng.clj:
```diff
@@ -101,55 +105,77 @@
+   (let [{:keys [scf-read-db url-prefix warn-experimental pretty-print]
+          :or {warn-experimental true
+               pretty-print false
+               url-prefix "/pdb"}} options
```
do we not have "/pdb" someplace as a top-level def?

kbarber replied:
Well, we have it defaulted in config, so on second thoughts I'll just stop defaulting it and make it mandatory. The decision to do this anyway was fairly arbitrary, it just helped 2-3 test cases, but not really necessary. I'm going to remove the defaulting and make it mandatory - since it did help detect a few cases I noticed where we weren't setting it.

senior commented on src/puppetlabs/puppetdb/query_eng/engine.clj:
```diff
@@ -1543,6 +1578,39 @@
+                          [] []
+                          :else (throw (IllegalArgumentException. "Your `from` query accepts an optional query only as a second argument. Check your query and try again.")))
+         entity (keyword (utils/underscores->dashes entity-str))]
+     (when warn
```
oh, warn-experimental is for experimental entities only? not for things like from?

kbarber replied:
It was created I believe to stop warnings just for those endpoints that we know PE uses regularly (at least for now). For the 'from' ... well, its experimental to everyone really, not just FOSS. I think its two different cases generally.

kbarber commented:
@ajroetker pretty=true works, do you mean its not in the docs?

senior commented:
:+1: just some minor comments

kbarber commented:
@senior suggested changes now up.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/319/
Test PASSed.

kbarber commented:
@senior @ajroetker green now, anything I've missed?
